src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,124 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_MAX_LENGTH * 2 + 1];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(chip::System::PacketBufferHandle::New(chip::System::kMaxPacketBufferSize));
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (enableRotatingDeviceId)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(lifetimeCounter, serialNumberBuffer, serialNumberBufferSize, rotatingDeviceIdBuffer,
+                                       ROTATING_DEVICE_ID_MAX_LENGTH * 2 + 1, rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(uint16_t lifetimeCounter, char * serialNumberBuffer,
+                                                                    size_t serialNumberBufferSize, char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdBufferOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Hash_SHA256_stream hash;
+    uint8_t outputBuffer[ROTATING_DEVICE_ID_MAX_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter outputBuferWriter(&outputBuffer[0], ArraySize(outputBuffer));
+    size_t rotatingDeviceIdBufferIndex = 0;
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= ROTATING_DEVICE_ID_MAX_LENGTH * 2 + 1, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Computing the Rotating Device Id
+    // RDI = Lifetime_Counter + SuffixBytes(HMAC_SHA256(Serial_Number + Lifetime_Counter), 16)
+
+    err = hash.Begin();
+    SuccessOrExit(err);
+
+    err = hash.AddData((uint8_t *) serialNumberBuffer, serialNumberBufferSize);
+    SuccessOrExit(err);
+    err = hash.AddData((uint8_t *) &lifetimeCounter, 1);
+    SuccessOrExit(err);
+    err = hash.Finish(hashOutputBuffer);
+    SuccessOrExit(err);
+
+    outputBuferWriter.Put16(lifetimeCounter);
+    outputBuferWriter.Put(
+        reinterpret_cast<const char *>(&hashOutputBuffer[kSHA256_Hash_Length - ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH]));
+
+    for (; rotatingDeviceIdBufferIndex < outputBuferWriter.Size(); rotatingDeviceIdBufferIndex++)
+    {
+        snprintf(&rotatingDeviceIdBuffer[rotatingDeviceIdBufferIndex * 2], 3, "%02X",
```

Comment:
it is not easy for me to guarantee at a glance that the given address always has 3 characters of space left. Could we update the code to make it more obvious? maybe not limit to 3 but rather

`sizeof(rotatingDeviceIdBuffer) - rotatingDeviceIdBufferIndex*2`

Reply:
sizeof causes a compilation error as it returns the size of char* , i used the size that's passed to the method.